src/server/pkg/hashtree/hashtree.go

```diff
@@ -56,6 +56,22 @@ type hashtree struct {
 	changed map[string]bool
 }
 
+// clone makes a deep copy of 'h' and returns it.
+func (h *hashtree) clone() (*hashtree, error) {
+	bts, err := h.Serialize()
+	if err != nil {
+		return nil, errorf(Internal, "could not clone HashTree: %s", err)
+	}
+	h2, err := Deserialize(bts)
+	if err != nil {
+		return nil, errorf(Internal, "could not clone HashTree: "+err.Error())
+	}
+	if h3, ok := h2.(*hashtree); ok {
+		return h3, nil
```

Comment:
Not that important but I think this is kind of an anti-pattern.  You should check for `!ok` and return the error in the if clause, and return `h3` in the main function body.